contentcuration/contentcuration/utils/channel.py

```diff
@@ -0,0 +1,142 @@
+import functools
+import math
+import random
+import time
+
+from django.core.cache import cache
+from django.db.models import Count
+from django.db.models import Sum
+from django_cte import With
+
+from contentcuration.models import Channel
+from contentcuration.models import ContentNode
+from contentcuration.models import FileCTE
+from contentcuration.models import User
+
+CACHE_CHANNEL_KEY = "channel_metadata_{}"
+
+
+def cache_stampede(expire, beta=1):
+    """Cache decorator with cache stampede protection.
+    Based on http://www.vldb.org/pvldb/vol8/p886-vattani.pdf (research by
+    Vattani, A.; Chierichetti, F.; Lowenstein, K. (2015), Optimal Probabilistic
+    Cache Stampede Prevention, VLDB, pp. 886-897, ISSN 2150-8097)
+    The cache stampede problem (also called dog-piling, cache miss storm,
+    or cache choking) is a situation that occurs when a popular cache item
+    expires, leading to multiple requests seeing a cache miss and
+    regenerating that same item at the same time
+
+    This  decorator implements cache stampede protection through
+    early recomputation. Early recomputation of function results will occur
+    probabilistically before expiration in a background thread of
+    execution.
+
+    If name is set to None (default), the callable name will be determined
+    automatically.
+    The original underlying function is accessible through the `__wrapped__`
+    attribute. This is useful for introspection, for bypassing the cache, or
+    for rewrapping the function with a different cache.
+    :param str key: key used to store the value in the cache
+    :param float expire: seconds until arguments expire
+    :param int beta: the parameter beta can be set to a value greater than 1 to
+                     favor earlier recomputations and further reduce stampedes but
+                     the paper authors show that setting beta=1 works well in practice
+    :return: callable decorator
+    """
+
+    def decorator(func):
+        def timer(*args, **kwargs):
+            "Time execution of `func` and return result and time delta."
+            start = time.time()
+            result = func(*args, **kwargs)
+            delta = time.time() - start
+            # The variable delta represents the time to recompute the value
+            # and is used to scale the probability distribution appropriately.
+            return result, delta
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
```

Comment:
If we want to reuse this decorator, guessing we want to be able to parameterize the cache key template - but I think everything else could stay the same (except `channel_id` would be a more generic id variable name). Thoughts?

Reply:
yes, because the key is fetched from the decorated function, the change should be calling it with the final key, applying the parametrization before, it should be an easy change. Do you want me to do it now, so it's prepare for it?

Reply:
Sure, if it's no bother!